lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+      }
+    });
+
+    // Note: 'operands' have already been converted with the supplied type
+    // converter to this pattern. Make sure that we materialize this conversion
+    // by updating the operands to op.
+    op->setOperands(operands);
+
+    return success(ok);
+  };
+  ArrayAttr parameters;
+};
+
+struct HWSpecializePass : public sv::HWSpecializeBase<HWSpecializePass> {
+  void runOnOperation() override;
+};
+
+static void populateTypeConversion(Location loc, TypeConverter &typeConverter,
+                                   ArrayAttr parameters) {
+  // Possibly parametric types
+  typeConverter.addConversion([=](hw::IntType type) {
+    return evaluateParametricType(loc, parameters, type).getValue();
+  });
+  typeConverter.addConversion([=](hw::ArrayType type) {
+    return evaluateParametricType(loc, parameters, type).getValue();
+  });
+
+  // Valid target types.
+  typeConverter.addConversion([](mlir::IntegerType type) { return type; });
+}
+
+// Specializes the provided 'base' module into the 'target' module. By doing
+// so, we create a new module which
+// 1. has no parameters
+// 2. has a name composing the name of 'base' as well as the 'parameters'
+// parameters.
+// 3. Has a top-level interface with any parametric types resolved.
+// 4. Any references to module parameters have been replaced with the
+// parameter value.
+static LogicalResult specializeModule(OpBuilder builder, ArrayAttr parameters,
+                                      HWModuleOp source, HWModuleOp &target) {
+  auto *ctx = builder.getContext();
+  // Update the types of the source module ports based on evaluating any
+  // parametric in/output ports.
+  auto ports = source.getPorts();
+  for (auto &in : llvm::enumerate(source.getFunctionType().getInputs())) {
+    auto resType =
+        evaluateParametricType(source.getLoc(), parameters, in.value());
+    if (failed(resType))
+      return failure();
+    ports.inputs[in.index()].type = resType.getValue();
+  }
+  for (auto &out : llvm::enumerate(source.getFunctionType().getResults())) {
+    auto resType =
+        evaluateParametricType(source.getLoc(), parameters, out.value());
+    if (failed(resType))
+      return failure();
+    ports.outputs[out.index()].type = resType.getValue();
+  }
+
+  // Create the specialized module using the evaluated port info.
+  target = builder.create<HWModuleOp>(
+      source.getLoc(),
+      StringAttr::get(ctx, generateModuleName(source, parameters)), ports);
+
+  // Erase the default created hw.output op - we'll copy the correct operation
+  // during body elaboration.
+  (*target.getOps<hw::OutputOp>().begin()).erase();
+
+  // Clone body of the source into the target.
+  BlockAndValueMapping mapper;
+  for (auto &&[src, dst] :
+       llvm::zip(source.getArguments(), target.getArguments()))
+    mapper.map(src, dst);
+  builder.setInsertionPointToStart(target.getBodyBlock());
+  for (auto &op : source.getOps())
```

Comment:
I don't think this handles backedges properly. IIRC, someone added something to CIRCT which does this clone correctly in the case of backedges. @mikeurbach would probably know.

Reply:
Hmm, are you referring to ValueMapper? I think Morten added that :) https://github.com/llvm/circt/blob/main/include/circt/Support/ValueMapper.h

Reply:
Good catch! I should know when to use my own code :facepalm: - will modify.